pyteal/stack_frame.py

```diff
@@ -0,0 +1,739 @@
+    via a NatalStackFrame object that is associated with the Expr object.
+
+    When source mapping is disabled (cf. `pyteal.ini`), NatalStackFrame'
+    constructor is a no-op.
+
+    When source mapping is enabled, it wraps a list of StackFrame's.
+
+    Under normal operations, only the "best" frame is kept in the list, so
+    the name is misleading.
+    """
+
+    _no_stackframes: bool = _sourcmapping_is_off()
+    _debug: bool = _debug_frames()
+    _keep_all_debugging = False
+
+    @classmethod
+    def sourcemapping_is_off(cls, _force_refresh: bool = False) -> bool:
+        """
+        The `_force_refresh` parameter, is mainly for test validation purposes.
+        It is discouraged for use in the wild because:
+        * Frames are useful in an "all or nothing" capacity. For example, in preparing
+            for a source mapping, it would be error prone to generate frames for
+            a subset of analyzed PyTeal
+        * Setting `_force_refresh = True` will cause a read from the file system every
+            time Frames are initialized and will result in significant performance degredation
+        """
+        if _force_refresh:
+            cls._no_stackframes = _sourcmapping_is_off()
+
+        return cls._no_stackframes
+
+    def __init__(
+        self,
+    ):
+        self._compiler_gen: bool = False
+        self._frames: list[StackFrame] = []
+
+        if self.sourcemapping_is_off():
+            return
+
+        # 1. get the full stack trace
+        full_stack = stack()
+
+        # 2. discard frames whose filename begins with "<"
+        frame_infos = list(filter(StackFrame._frame_info_not_py_crud, full_stack))
+
+        def _make_stack_frames(fis):
+            return [
+                frame
+                for f in fis
+                if (frame := StackFrame._init_or_drop(self, f, full_stack))
+            ]
+
+        if self._keep_all_debugging or len(frame_infos) <= 1:
+            self._frames = _make_stack_frames(frame_infos)
+            return
+
+        # 3. start the best frame search right after where NatalStackFrame() was constructed
+        # For more details see the unit test:
+        # tests/unit/sourcemap_monkey_unit_test.py::test_frame_info_is_right_before_core_last_drop_idx
+        i = 2  # formerly this was `last_drop_idx = 1; i = last_drop_idx + 1`
+
+        # 4. fast forward the right bound until we're out of pyteal-library code
+        # This sets last_keep_idx to the first frame index which isn't pyteal
+        while i < len(frame_infos) and StackFrame._frame_info_is_pyteal(frame_infos[i]):
+            i += 1
+        last_keep_idx = i
+
+        # 5. if the pyteal-library exit point was an import, the expression was
+        # generated by pyteal itself. So let's back up and look for a "# T2PT*" comment
+        # which will give us a clue for what to do with this expression
+        if StackFrame._frame_info_is_pyteal_import(frame_infos[last_keep_idx]):
```

Comment:
at this point, seems like the thing in `frame_infos[last_keep_idx]` is:
- imported by pyteal
- yet it is build from pyteal lib

Reply:
Yes, but I have a bit more streamlined criteria coming up in #676.